datastruct/mpool.h

```diff
@@ -11,69 +13,114 @@
  * of allowing the memory to be reused by some other code.
  */
 
+struct mpool {
+	size_t stacklen;
+	size_t allocsize;
+	void ** allocs;
+	uint64_t nallocs;
+	uint64_t nempties;
+	int state;
+	void ** allocs_static;
+	void (*atexitfunc)(void);
+};
+
+static inline void
+mpool_atexit(struct mpool * M)
+{
+
+	while (M->stacklen)
+		free(M->allocs[--M->stacklen]);
+	if (M->allocs != M->allocs_static)
+		free(M->allocs);
+}
+
+static inline void *
+mpool_malloc(struct mpool * M, size_t len)
+{
+
+	M->nallocs++;
+	if (M->stacklen)
+		return (M->allocs[--(M->stacklen)]);
```

Comment:
Also, shouldn't we be returning `&M->allocs[--(M->stacklen)]);` here?  Because the initial contents of `static void * mpool_##name##_static` will be undefined.

Reply:
No, because we're only going to return that value after it has been written (and stacklen has been incremented).

Also, M->allocs holds pointers to memory allocations.  Those pointers are the ones we want to return; we don't want to return a pointer to a pointer.